### PR TITLE
Add caching of CDS responses

### PR DESCRIPTION
**Pull request by howardjohn**
I have placed a hold to allow all 3 "assignees" to review (and anyone interested) before merging, as this is a fairly complex change

Before:

After:

Cache hit rate is 95%+:

A single cluster is ~900B typically. With 60k max entries, this results in at most 60mb overhead.

Istiod is using ~6GB memory for this case so that is only 1% extra memory usage, well worth the CPU gains

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
/retest

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:194]: Where were we doing this before? If we weren't why do we need it for caching?

**Review by stevenctl — COMMENTED** [review]
Looks good!

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:194]: We need to apply the patches on a per-cluster basis now, where-as the old way (conditionallyAppend) did many. I also thing there may be some merge conflict here at some point, though so I probably need to cxlean it up

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:129]: nit: can we move this to a constructor?

File: releasenotes/notes/cds-cache.yaml
Comment (stevenctl) [inline releasenotes/notes/cds-cache.yaml:8]:
```suggestion
  **Improved** the XDS configuration generation to cache Envoy clusters, substantially improving performance of the control plane in many cases
```

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:116]: Where are these used? Intended to emit metrics?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:142]: Because of how tightly coupled the default cluster is with its subsets, this feels more like we are scaling the hit/miss rates by the "size" of the set of clusters. 

Not sure if it actually makes more sense to count the "set" of clusters as defined by host/port though..

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (stevenctl) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:193]: In that case can we dry up conditionallyAppend by calling this there?

**Review by stevenctl — COMMENTED** [review]
mostly nits, this is great!

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:129]: I would be slightly worried about creating a func with tons of parameters. At this point I think we just need `cb`,, service, and port though so perhaps its ok, will try it out

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:116]: Ah yeah, meant to return something like https://github.com/istio/istio/blob/cd4e73cd8fdca373d89ae1b8175a1b0c2d4d0658/pilot/pkg/xds/eds.go#L426 but forgot to finish it out.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:142]: I see what you mean, one thing I like about this is when I fix the logging to properly include the caching we will see something like `CDS: push resources:18 cached:7/18`. If we do it by "set", we may see  `CDS: push resources:18 cached:3/7` or something, which may be a bit confusing?

Not a strong preference though

**Comment by hzxuzhonghu** [comment]
wait, this is on my todo list

**Comment by howardjohn** [comment]
Yeah I will wait until all 3 on the list have had a chance to review since
its a big change

File: pilot/pkg/networking/core/configgen.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/configgen.go:38]: why do we need to return `model.XdsLogDetails`? I wonder if theres a better, cleaner way to do this

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:174]: I don't really understand why we increment by the length of all the clusters

shouldn't we only increment `hit` or `miss` by the number of clusters that were a "hit" or "miss"? why does allfound matter?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:193]: comment might be useful

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:200]: comment?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:251]:
```suggestion
```

cache is enabled, right?

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder.go:504]:
```suggestion
	locs := lbEndpoints
```
is there a semantic difference?

**Review by adiprerepa — COMMENTED** [review]
basically lgtm -- does it make sense to add tests to this?

File: pilot/pkg/networking/core/configgen.go
Comment (howardjohn) [reply pilot/pkg/networking/core/configgen.go:38]: To bubble up the cache hit/miss details so we can include it in the logs.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:174]: the length of all the clusters is == the number of hit or miss. They either all hit or they all miss because we require all subsets to be a hit for it to be considered a hit

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:251]: Not for this part of the code. Only for buildOutboundClusters

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:504]: Yes, one creates an empty list and one makes an alias to a map. We need the list to retain sorted order for deterministic XDS output

File: pilot/pkg/features/pilot.go
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:348]: I think so. This is mostly an escape hatch if something went horribly wrong so users can turn it off without reverting their Istio update. Assuming no bugs, there is no downsides to this change

**Comment by howardjohn** [comment]
> basically lgtm -- does it make sense to add tests to this?

We do have one almost very good test - UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS. This checks a bunch of invariants of the cache. Unfortunately we do not run it all the time since it changes things quite a bit and is extremely slow. However, I ran it a bunch on this PR prior to marking it ready and it did not fail.

I have a pretty high level of confidence with UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS - it found 5+ extremely obscure bugs in the EDS cache. As a result, I think most of the caching issues have already been addressed (hopefully).

I am definitely open to suggestions on how to test it better though!

**Comment by adiprerepa** [comment]
> I have a pretty high level of confidence with UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS - it found 5+ extremely obscure bugs in the EDS cache. As a result, I think most of the caching issues have already been addressed (hopefully).

My main concern is the future, where if someone introduces a bug into the cache, and we don't enable `UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS` every time, we may not catch it. However, I don't really have any suggestions, so I'll defer to your judgement.

**Comment by howardjohn** [comment]
> My main concern is the future, where if someone introduces a bug into the cache, and we don't enable UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS every time, we may not catch it. However, I don't really have any suggestions, so I'll defer to your judgement.
I have the same concern. Not sure a great solution, maybe we start running a dedicated job with that enabled in postsubmit

File: pilot/pkg/features/pilot.go
Comment (ramaraochavali) [reply pilot/pkg/features/pilot.go:348]: +1 for enabling it by default.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:178]: nit: Can you add a comment here that we did not find cluster in cache so building now?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:112]: resources here contain cached as well as non cached clusters correct? Are n't we caching Any already? If yes, we do not have to call util.MessageToAny() for cached clusters?

**Comment by ramaraochavali** [comment]
generally looks good. Added some comments. But would like to have another detailed review.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:112]: `resources` contains already marshalled to Any. This is returned by outbound cluster builder because we have to cache the any.

The other builders which are not yet doing caching are just returning the original Cluster, so now we marshal them.

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:340]: We *can* but it's quite hard. I can invest some efforts into it now but I was hoping that we could merge the initial form and then optimize it in the near future.

I completely agree it's not ideal but it's actually not as bad as it seems - if you look in the pr description in my tests we were hitting 95% cache hit rate, so it ends up being fairly efficient

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:197]: It needs to be until your PR merged, right?

I would like to keep the 2 PRs decoupled since they are not dependant on each other. That way either one can merge first

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:118]: It seems useful to me personally. We do the same for EDS which has a lot higher volume and it hasn't shown up on any profiles. I am open to removing but prefer to keep things debugable unless proven to cause performance issues

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:251]: There is none other than trying to keep a big PR from being even bigger.

The sni dnat is only for a very small number of proxies so it's not a high priority at all. It may not even be effective to cache given it will take up space in the cache but is unlikely to cause cache hit

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:197]: make sense

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:251]: I guessed this reason

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:340]: I am worried this can cause mem leak if the cache size is very large as i don't see cache clear when push version changed.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:112]: So the outbound clusters are marshalled twice - I do not know the impact of it, but is it possible to avoid it like adding them to cachedresources and add cachedresources as is at the end?

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:340]: A single large cluster uses 1kb. The cache has a hard limit of 60k entries. This leads to an upper bound of 60mb of memory used by the cache.

I do agree we do not clear it though. I will need to think about how we can address that. However, the leak does have a strict upper bound so its probably still better to have this with leak than not at all

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:112]: They are not marshalled twice.

For outbound (with caching): we return them as `Any` and add to `resources`.
For other types (without caching): we return them as `Cluster`, add them to `clusters`, then marshal and add to `resources`.

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:340]: I am thinking more about this, could we remove the dependency on push version, and add envoyfilters here?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:112]: Makes Sense. Sorry I missed in the code that outbound clusters are being directly added to resources.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:86]: Only for now, in the future it will be needed once other methods add cache

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:340]: I will do it in a followup, I really do not want to tie this PR to that change as its complicated enough. I am fine with disabling by default until then if you want.

**Review by stevenctl — APPROVED** [review]

File: pilot/pkg/xds/bench_test.go
Comment (adiprerepa) [inline pilot/pkg/xds/bench_test.go:114]: what happened to the log logic?

**Review by adiprerepa — COMMENTED** [review]
lgtm

File: pilot/pkg/xds/bench_test.go
Comment (howardjohn) [reply pilot/pkg/xds/bench_test.go:114]: added back, good catch

**Review by hzxuzhonghu — COMMENTED** [review]
lgtm

**Pull request merged**
